src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2812,13 +2821,71 @@ void EmitAtomicSingleCharLoop(RegexNode node)
                 ReturnInt32Local(iterationLocal);
             }
 
+            // Emits the code to handle a non-backtracking optional zero-or-one loop.
+            void EmitAtomicSingleCharZeroOrOne(RegexNode node)
```

Comment:
It's a shame that RegexNode/Tree leaks to RegexCompiler now. Could some of these optimizations be done in RegexWriter or earlier, such that RegexCompiler only needs to work off the RegexCode?

Also are any of these relevant to the interpreter?

Reply:
> It's a shame that RegexNode/Tree leaks to RegexCompiler now.

That's not new to this change.  The entire non-backtracking compiler implementation works off of the regex tree rather than off of the regex code.  Doing this in the regex code would require overhauling it to create an entirely new byte code that effectively maps to the regex tree.

> Also are any of these relevant to the interpreter?

No.  This PR is about how we generate the code in the non-backtracking compiler for a ? quantifier.